### PR TITLE
[XR] Input transformation should be applied during the XR loop

### LINKED ISSUE #14016: [XR] Node parented to the XR camera is incorrectly positioned
See https://forum.babylonjs.com/t/child-object-of-xr-camera-shake-in-some-cases/42043

### PR DESCRIPTION
**Pull request by RaananW**
When using the standard input system (i.e. changing cameraDirection and cameraRotation) the transformation changes instantly, but in XR, due to the way we change the camera transformation (during the XR loop) this has to be applied deferred.

The initial solution was to change the XR camera only, but this required repeating a lot of code because of the inheritence architecture.

This way we maintain back-compat while allowing "deferred" updating the camera's transformation based on inputs.

Fixes #14016
PG to test - #m1kgcr

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

File: packages/dev/core/src/Cameras/freeCamera.ts [hunk 1]
```diff
@@ -376,36 +376,47 @@ export class FreeCamera extends TargetCamera {
...
         }
 
-        coordinator.getNewPosition(this._oldPosition, actualDisplacement, this._collider, 3, null, this._onCollisionPositionChange, this.uniqueId);
+        coordinator.getNewPosition(
+            this._oldPosition,
+            actualDisplacement,
+            this._collider,
+            3,
+            null,
+            (_collisionId: number, newPosition: Vector3, collidedMesh: Nullable<AbstractMesh> = null) => {
```
Comment (sebavan) [inline packages/dev/core/src/Cameras/freeCamera.ts:385]: Isnt this creating more GC ?

File: packages/dev/core/src/Cameras/freeCamera.ts [hunk 2]
```diff
@@ -376,36 +376,47 @@ export class FreeCamera extends TargetCamera {
...
+                const updatePosition = (newPos: Vector3) => {
```
Comment (sebavan) [inline packages/dev/core/src/Cameras/freeCamera.ts:386]: Here as well, not sure to understand why we need to define updatePosition and use it here only once ?

File: packages/dev/core/src/Cameras/targetCamera.ts [hunk 3]
```diff
@@ -313,54 +318,88 @@ export class TargetCamera extends Camera {
...
 
         // Move
         if (needToMove) {
-            this._updatePosition();
+            this._updatePosition(defferOnly);
         }
 
         // Rotate
         if (needToRotate) {
+            console.log("rotate");
```
Comment (sebavan) [inline packages/dev/core/src/Cameras/targetCamera.ts:361]: to be removed

**Review by sebavan — COMMENTED** [review]
as it seems defer would ever only be for XR, I wonder if there is a way to limit the impact in the target camera with the extra copies ?

**Comment by RaananW** [comment]
> as it seems defer would ever only be for XR, I wonder if there is a way to limit the impact in the target camera with the extra copies ?

I just noticed I didn't open that as a draft. Was supposed to be, because this is just a suggestion. the inheritance between the classes forces us to do it this way, or reimplement the entire thing in the WebXR camera (which will mean repeating the code).

The first version I implemented did exactly that (you can see the commit history), but I decided to prevent the reimplementation.
I agree that this is only for the XR camera. the question is, do we want to reimplement (and avoid using inheritance) and risk future issues with changes to the parent classes, or stay OOP-oriented.

File: packages/dev/core/src/Cameras/freeCamera.ts (hunk 2 above)
Comment (RaananW) [reply packages/dev/core/src/Cameras/freeCamera.ts:386]: This code existed already, I didn't change it. I agree it is not needed :-) 
I can fix that

File: packages/dev/core/src/Cameras/freeCamera.ts (hunk 1 above)
Comment (RaananW) [reply packages/dev/core/src/Cameras/freeCamera.ts:385]: There is no way to pass the boolean to the function, if used as a callback. That would require changing the collider's signature as well.

File: packages/dev/core/src/Cameras/camera.ts [hunk 4]
```diff
@@ -685,7 +685,7 @@ export class Camera extends Node {
     }
 
     /** @internal */
-    public _checkInputs(): void {
+    public _checkInputs(_defferOnly?: boolean): void {
```
Comment (PolygonalSun) [inline packages/dev/core/src/Cameras/camera.ts:688]: Nit: `_deferOnly`

File: packages/dev/core/src/Cameras/freeCamera.ts (hunk 1 above)
Comment (sebavan) [reply packages/dev/core/src/Cameras/freeCamera.ts:385]: the mode could be a property of the class instead ? as it is always the same per camera type ?

Comment (RaananW) [reply packages/dev/core/src/Cameras/freeCamera.ts:385]: sure, it can. can be a flag on the class level instead of being passed as is it now

Comment (RaananW) [reply packages/dev/core/src/Cameras/freeCamera.ts]: Done :-)

**Review by sebavan — COMMENTED** [review]
Also it seems onCollide is not called if deffered anymore ?

File: packages/dev/core/src/XR/webXRCamera.ts [hunk 5]
```diff
@@ -79,6 +80,12 @@ export class WebXRCamera extends FreeCamera {
                 if (this._firstFrame) {
                     this._updateFromXRSession();
                 }
+
+                if (this._deferredUpdated) {
+                    this.position.copyFrom(this._deferredPositionUpdate);
+                    this.rotationQuaternion.copyFrom(this._deferredRotationQuaternionUpdate);
```
[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/XR/webXRCamera.ts:86]: No, rotation is not being used if rotationQuaternion is.

**Comment by RaananW** [comment]
> Also it seems onCollide is not called if deffered anymore ?

onCollide is an interesting task in case of deferred position change :-) 
What i ended up doing is call onCollide on time. position might change in the next frame (depending on the XR vs. render loop)

File: packages/dev/core/src/XR/webXRCamera.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (sebavan) [reply packages/dev/core/src/XR/webXRCamera.ts:86]: isnt the rotation used line 349 as the bases for rotation on the next check ?

[the remark this answers is not in the document]
Comment (RaananW) [reply packages/dev/core/src/XR/webXRCamera.ts:86]: it is copied from rotationQuaternion, if available (in line 351)

**Review by sebavan — APPROVED** [review]

**Pull request merged**